addons/account/models/account_bank_statement.py

```diff
@@ -407,15 +407,26 @@ def _needaction_domain_get(self):
     def button_cancel_reconciliation(self):
         # TOCKECK : might not behave as expected in case of reconciliations (match statement line with already
         # registered payment) or partial reconciliations : it will completely remove the existing payment.
-        move_recs = self.env['account.move']
+        move_recs = moves_with_payment = moves_without_payment = self.env['account.move']
         for st_line in self:
             move_recs = (move_recs | st_line.journal_entry_ids)
         if move_recs:
             for move in move_recs:
-                move.line_ids.remove_move_reconcile()
+                if any(line.payment_id for line in move.line_ids):
+                    moves_with_payment = (moves_with_payment | move)
+                else:
+                    moves_without_payment = (moves_without_payment | move)
+                if move in moves_with_payment:
```

Comment:
it's a bit strange as a a structure... I'd put only one if...else statement. Something like 
```
+                if any(line.payment_id for line in move.line_ids):
+                    moves_with_payment = (moves_with_payment | move)
+                    for line in move.line_ids:
+                        if line.statement_id in self.statement_id:
+                            line.statement_id = False
+                        if line.payment_id and line.payment_id.state == 'reconciled':
+                            line.payment_id.state = 'posted'
+                else:
+                    moves_without_payment = (moves_without_payment | move)
+                    move.line_ids.remove_move_reconcile()
```
and the sets moves_without_payment and moves_with_payment may eventually become useless (button_cancel() and unlink() could be called directly in the loop without too much perf harm I supose)